sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/UnboundedReaderImpl.java

```diff
@@ -0,0 +1,136 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.pubsublite.internal;
+
+import static org.apache.beam.sdk.io.gcp.pubsublite.internal.ApiServices.asCloseable;
+import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkArgument;
+import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkState;
+
+import com.google.api.core.ApiService.State;
+import com.google.cloud.pubsublite.Offset;
+import com.google.cloud.pubsublite.proto.SequencedMessage;
+import com.google.protobuf.util.Timestamps;
+import java.io.IOException;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import org.apache.beam.sdk.io.UnboundedSource;
+import org.apache.beam.sdk.io.UnboundedSource.UnboundedReader;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.joda.time.Instant;
+
+public class UnboundedReaderImpl extends UnboundedReader<SequencedMessage> {
+
+  private final UnboundedSource<SequencedMessage, CheckpointMarkImpl> source;
+  private final MemoryBufferedSubscriber subscriber;
+  private final TopicBacklogReader backlogReader;
+  private final BlockingCommitter committer;
+
+  private Offset fetchOffset;
+  private Optional<Instant> lastMessageTimestamp = Optional.empty();
+  private boolean advanced = false;
+
+  UnboundedReaderImpl(
+      UnboundedSource<SequencedMessage, CheckpointMarkImpl> source,
+      MemoryBufferedSubscriber subscriber,
+      TopicBacklogReader backlogReader,
+      BlockingCommitter committer,
+      Offset initialOffset) {
+    checkArgument(initialOffset.equals(subscriber.fetchOffset()));
+    this.source = source;
+    this.subscriber = subscriber;
+    this.backlogReader = backlogReader;
+    this.committer = committer;
+    this.fetchOffset = initialOffset;
+  }
+
+  @Override
+  public SequencedMessage getCurrent() throws NoSuchElementException {
+    if (!advanced) {
+      throw new NoSuchElementException();
+    }
+    return subscriber.peek().get();
+  }
+
+  @Override
+  public Instant getCurrentTimestamp() throws NoSuchElementException {
+    return getTimestamp(getCurrent());
+  }
+
+  private static Instant getTimestamp(SequencedMessage message) {
+    return Instant.ofEpochMilli(Timestamps.toMillis(message.getPublishTime()));
+  }
+
+  @Override
+  public void close() throws IOException {
+    try (AutoCloseable c1 = backlogReader;
+        AutoCloseable c2 = committer;
+        AutoCloseable c3 = asCloseable(subscriber)) {
+    } catch (Exception e) {
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public boolean start() throws IOException {
+    try {
+      subscriber.startAsync().awaitRunning(1, TimeUnit.MINUTES);
+    } catch (Exception e) {
+      throw new IOException(e);
+    }
+    return advance();
```

Comment:
Probably check subscriber state and fail if not State.RUNNING ? Otherwise if the reader doesn't get to the correct state here (State.RUNNING) the reader will never be able to recover.

Reply:
This is the first line of advance(), so this check is already made.